In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler, MinMaxScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
import matplotlib.pyplot as plt
import seaborn as sns

PATH = "feature_engineered/feature_engineered_CMG_RRV.csv"
df = pd.read_csv(PATH)

In [2]:
len(df.columns)

155

In [3]:
df.columns.tolist()

['date',
 'close',
 'open',
 'high',
 'low',
 'volume',
 'change',
 'symbol',
 'return_day',
 'return_week',
 'return_month',
 'volatility_day',
 'volatility_week',
 'volatility_month',
 'liquidity_day',
 'liquidity_week',
 'liquidity_month',
 'high_minus_close',
 'low_minus_open',
 'cumulative_return',
 'stochastic_osc',
 'atr',
 'sma_3',
 'sma_7',
 'sma_14',
 'sma_21',
 'sma_50',
 'sma_100',
 'wma_3',
 'wma_7',
 'wma_14',
 'wma_21',
 'wma_50',
 'wma_100',
 'ema_6',
 'ema_12',
 'ema_26',
 'out_macd',
 'out_macd_signal',
 'out_macd_hist',
 'rsi_6',
 'rsi_12',
 'rsi_14',
 'stochrsi_6',
 'stochrsi_12',
 'stochrsi_14',
 'bbands_middle',
 'bbands_upper',
 'bbands_lower',
 'obv',
 'mfi_14',
 'mom_1',
 'mom_3',
 'mom_7',
 'cci_12',
 'cci_20',
 'rocr_3',
 'rocr_12',
 'willr',
 'trix',
 'net_profit_loss_before_tax',
 'depreciation_fixed_assets',
 'credit_risk_reserve',
 'unrealized_forex_gain_loss',
 'investment_income_loss',
 'interest_income',
 'interest_dividend_income',
 'net_cash_flow_ope

In [12]:
features = [
#  'date',
 'close',
 'open',
 'high',
 'low',
 'volume',
 'change',
#  'symbol',
 'return_day',
 'return_week',
 'return_month',
 'volatility_day',
 'volatility_week',
 'volatility_month',
 'liquidity_day',
 'liquidity_week',
 'liquidity_month',
 'high_minus_close',
 'low_minus_open',
 'cumulative_return',
 'stochastic_osc',
 'atr',
 'sma_3',
 'sma_7',
 'sma_14',
 'sma_21',
 'sma_50',
 'sma_100',
 'wma_3',
 'wma_7',
 'wma_14',
 'wma_21',
 'wma_50',
 'wma_100',
 'ema_6',
 'ema_12',
 'ema_26',
 'out_macd',
 'out_macd_signal',
 'out_macd_hist',
 'rsi_6',
 'rsi_12',
 'rsi_14',
 'stochrsi_6',
 'stochrsi_12',
 'stochrsi_14',
 'bbands_middle',
 'bbands_upper',
 'bbands_lower',
 'obv',
 'mfi_14',
 'mom_1',
 'mom_3',
 'mom_7',
 'cci_12',
 'cci_20',
 'rocr_3',
 'rocr_12',
 'willr',
 'trix',
 'net_profit_loss_before_tax',
 'depreciation_fixed_assets',
 'credit_risk_reserve',
 'unrealized_forex_gain_loss',
 'investment_income_loss',
 'interest_income',
 'interest_dividend_income',
 'net_cash_flow_operating_before_wc_changes',
 'change_in_receivables',
 'change_in_inventories',
 'change_in_payables',
 'change_in_prepaid_expenses',
 'interest_expense_paid',
 'corporate_income_tax_paid',
 'other_cash_inflows_operating',
 'other_cash_outflows_operating',
 'net_cash_flow_from_business_activities',
 'capital_expenditure',
 'cash_from_fixed_assets_sale',
 'cash_outflow_loans_debt_instruments',
 'cash_inflow_loan_recoveries',
 'investments_in_other_businesses',
 'cash_from_sale_investments',
 'dividends_profit_received',
 'cash_flow_investing',
 'increase_in_equity',
 'cash_used_share_repurchase',
 'cash_from_borrowings',
 'cash_repayment_borrowings',
 'dividends_paid',
 'cash_flow_financing',
 'net_cash_flow_period',
 'cash_and_equivalents',
 'exchange_rate_effect',
 'ending_cash_equivalents',
 'revenue_growth_percent',
 'revenue_vnd',
 'net_profit_after_tax_parent_vnd',
 'profit_growth_percent',
 'financial_income',
 'interest_expense',
 'sales_revenue_service_income',
 'revenue_deductions',
 'net_revenue',
 'cogs',
 'gross_profit',
 'financial_expenses',
 'profit_loss_associates',
 'selling_expenses',
 'administrative_expenses',
 'operating_profit_loss',
 'other_income',
 'profit_loss_joint_ventures',
 'other_income_expense',
 'other_profit',
 'profit_before_tax',
 'current_corporate_tax_expense',
 'deferred_corporate_tax_expense',
 'net_profit',
 'minority_interest',
 'parent_company_shareholders',
 'loans_to_equity_ratio',
 'debt_to_equity',
 'fixed_assets_to_equity',
 'equity_to_chartered_capital',
 'asset_turnover',
 'fixed_asset_turnover',
 'avg_collection_period',
 'avg_inventory_days',
 'avg_payment_days',
 'cash_cycle',
 'inventory_turnover',
 'ebit_margin_percent',
 'gross_profit_margin_percent',
 'net_profit_margin_percent',
 'roe_percent',
 'roic_percent',
 'roa_percent',
 'ebitda_billion_vnd',
 'ebit_billion_vnd',
 'dividend_yield_percent',
 'current_ratio',
 'cash_ratio',
 'quick_ratio',
 'interest_coverage',
 'financial_leverage',
 'market_cap_billion_vnd',
 'shares_outstanding_million',
 'pe',
 'pb',
 'ps',
 'p_cash_flow',
 'eps_vnd',
 'bvps_vnd',
 'ev_ebitda',
]

In [5]:
df.isna().sum()[df.isna().sum() > 0]

Series([], dtype: int64)

In [6]:
df.head()

,date,close,open,high,low,volume,change,symbol,return_day,return_week,...,financial_leverage,market_cap_billion_vnd,shares_outstanding_million,pe,pb,ps,p_cash_flow,eps_vnd,bvps_vnd,ev_ebitda
0,2018-03-31,20865.1,20259.533333,21162.4,19951.233333,321393.333333,3.50,CMG_RRV,-0.000527,0.165433,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527
1,2018-04-01,20854.1,20699.966667,21151.4,20083.366667,246696.666667,1.67,CMG_RRV,-0.000527,0.127377,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527
2,2018-04-02,20843.1,21140.400000,21140.4,20215.500000,172000.000000,-0.16,CMG_RRV,-0.000527,0.099302,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527
3,2018-04-03,20479.8,20545.800000,20810.1,20413.700000,165830.000000,-1.74,CMG_RRV,-0.017430,0.033337,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527
4,2018-04-04,20744.0,20479.800000,21140.4,20479.800000,288310.000000,1.29,CMG_RRV,0.012901,-0.006328,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527


# Feature Selection

In [ ]:
import pandas as pd
import numpy as np
from sklearn.feature_selection import mutual_info_regression, f_regression
import seaborn as sns
import matplotlib.pyplot as plt

target = 'close'  

X = df.select_dtypes(include=[np.number]).drop(columns=[target])
y = df[target]

mi_scores = mutual_info_regression(X, y)
mi_series = pd.Series(mi_scores, index=X.columns)

_, p_values = f_regression(X, y)
p_series = pd.Series(p_values, index=X.columns)

selected_features = mi_series[(p_series < 0.05)].sort_values(ascending=False)

# 4. Draw correlation matrix of selected features
top_features = selected_features.head(15).index  # You can adjust this
corr_matrix = df[top_features].corr()

# Plot correlation matrix
plt.figure(figsize=(12, 8))
sns.heatmap(corr_matrix, annot=True, fmt=".2f", cmap="coolwarm")
plt.title("Correlation Matrix of Selected Features")
plt.show()

Standardization and normalization

In [13]:
data = df[features].values

# scaler = StandardScaler()
scaler = MinMaxScaler()
data_scaled = scaler.fit_transform(data)

In [14]:
len(features)

153

In [15]:
df.head()

,date,close,open,high,low,volume,change,symbol,return_day,return_week,...,financial_leverage,market_cap_billion_vnd,shares_outstanding_million,pe,pb,ps,p_cash_flow,eps_vnd,bvps_vnd,ev_ebitda
0,2018-03-31,20865.1,20259.533333,21162.4,19951.233333,321393.333333,3.50,CMG_RRV,-0.000527,0.165433,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527
1,2018-04-01,20854.1,20699.966667,21151.4,20083.366667,246696.666667,1.67,CMG_RRV,-0.000527,0.127377,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527
2,2018-04-02,20843.1,21140.400000,21140.4,20215.500000,172000.000000,-0.16,CMG_RRV,-0.000527,0.099302,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527
3,2018-04-03,20479.8,20545.800000,20810.1,20413.700000,165830.000000,-1.74,CMG_RRV,-0.017430,0.033337,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527
4,2018-04-04,20744.0,20479.800000,21140.4,20479.800000,288310.000000,1.29,CMG_RRV,0.012901,-0.006328,...,2.4009,1380510036500,67341953,8.909423,1.414769,0.283481,6.86447,69.83477,14490.001038,5.024527


In [16]:
df.tail()

,date,close,open,high,low,volume,change,symbol,return_day,return_week,...,financial_leverage,market_cap_billion_vnd,shares_outstanding_million,pe,pb,ps,p_cash_flow,eps_vnd,bvps_vnd,ev_ebitda
2463,2024-12-27,48559.000000,49730.000000,49910.000000,48468.000000,3.600000e+06,-2.710000,CMG_RRV,-0.027069,-0.038061,...,2.227640,9468086700800,211341221,29.761275,3.458975,1.222675,14.522856,278.574989,12951.813926,14.864212
2464,2024-12-28,48739.333333,49486.666667,49773.333333,48645.333333,2.721233e+06,-1.436667,CMG_RRV,0.003714,-0.030449,...,2.227640,9468086700800,211341221,29.761275,3.458975,1.222675,14.522856,278.574989,12951.813926,14.864212
2465,2024-12-29,48919.666667,49243.333333,49636.666667,48822.666667,1.842467e+06,-0.163333,CMG_RRV,0.003700,-0.012721,...,2.227640,9468086700800,211341221,29.761275,3.458975,1.222675,14.522856,278.574989,12951.813926,14.864212
2466,2024-12-30,49100.000000,49000.000000,49500.000000,49000.000000,9.637000e+05,1.110000,CMG_RRV,0.003686,-0.025020,...,2.227640,9468086700800,211341221,29.761275,3.458975,1.222675,14.522856,278.574989,12951.813926,14.864212
2467,2024-12-31,49250.000000,49300.000000,49850.000000,48950.000000,9.573000e+05,0.310000,CMG_RRV,0.003055,-0.013224,...,2.152912,8695028208050,211300807,27.838678,3.047378,1.095932,11.593105,599.010040,13503.411529,12.007542


In [17]:
len(df)

2468

In [18]:
window_size = 30
forecast_horizon = 30

def create_sequences(data, window_size, forecast_horizon):
    X, y = [], []
    for i in range(len(data) - window_size - forecast_horizon + 1):
        X.append(data[i : i + window_size])
        y.append(data[i + window_size : i + window_size + forecast_horizon, 0])
    return np.array(X), np.array(y)

X, y = create_sequences(data_scaled, window_size, forecast_horizon)
print("X shape:", X.shape)  # Kích thước: (số mẫu, window_size, số đặc trưng)
print("y shape:", y.shape)  # Kích thước: (số mẫu, forecast_horizon)

X shape: (2409, 30, 153)
y shape: (2409, 30)


With 153 features, you may have redundant or noisy information. It can be beneficial to first explore feature selection or dimensionality reduction (e.g., PCA) rather than simply increasing model complexity.

In [19]:
model = Sequential()
model.add(LSTM(100, activation='relu', return_sequences=True, input_shape=(window_size, X.shape[2])))
model.add(Dropout(0.2))  

model.add(LSTM(50, activation='relu'))
model.add(Dropout(0.2))  
model.add(Dense(32, activation='relu'))
model.add(Dropout(0.2))  
model.add(Dense(forecast_horizon))
model.compile(optimizer='adam', loss='mse')
model.summary()

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/keras/src/layers/rnn/rnn.py:200: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ lstm (LSTM)                     │ (None, 30, 100)        │       101,600 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 30, 100)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_1 (LSTM)                   │ (None, 50)             │        30,200 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 50)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 32)             │         1,632 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 32)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 30)             │           990 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 134,422 (525.09 KB)

 Trainable params: 134,422 (525.09 KB)

 Non-trainable params: 0 (0.00 B)

In [20]:
# model training
history = model.fit(X, y, epochs=20, batch_size=64, validation_split=0.2)

Epoch 1/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 2s 34ms/step - loss: 0.1437 - val_loss: 0.1251
Epoch 2/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 28ms/step - loss: 0.0533 - val_loss: 0.0590
Epoch 3/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 29ms/step - loss: 0.0269 - val_loss: 0.0305
Epoch 4/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 30ms/step - loss: 0.0179 - val_loss: 0.0357
Epoch 5/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 31ms/step - loss: 0.0122 - val_loss: 0.0238
Epoch 6/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 30ms/step - loss: 0.0102 - val_loss: 0.0270
Epoch 7/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 29ms/step - loss: 0.0084 - val_loss: 0.0229
Epoch 8/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 30ms/step - loss: 0.0085 - val_loss: 0.0418
Epoch 9/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 30ms/step - loss: 0.0069 - val_loss: 0.0376
Epoch 10/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 28ms/step - loss: 0.0068 - val_loss: 0.0338
Epoch 11/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 36ms/step - loss: 0.0066 - val_loss: 0.0343
Epoch 12/20
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 31ms/step - loss: 0.0

In [ ]:
# Prediction for the next 30 days
last_window = data_scaled[-window_size:]
last_window = np.expand_dims(last_window, axis=0) 

In [ ]:
# Predict using the model
predicted_scaled = model.predict(last_window)  # Shape: (1, forecast_horizon)
predicted_scaled.shape

In [ ]:
# Create dummy array to inverse transform (assume 'Price' is at index 0)
dummy = np.zeros((forecast_horizon, len(features)-1))
pred_full = np.concatenate([predicted_scaled.T, dummy], axis=1)
predicted_prices = scaler.inverse_transform(pred_full)[:, 0]  # Extract the 'Price' column

print("Dự báo giá trong {} ngày tới cho mã là:".format(forecast_horizon))
print(predicted_prices)

In [ ]:
daily_file = "processed_stck_data/CMG.csv"
df_daily = pd.read_csv(daily_file, parse_dates=['Date'])
df_daily.sort_values("Date", inplace=True)
last_date_str = "2024-12-31"
last_date = pd.to_datetime(last_date_str)

In [ ]:
# Training cutoff was 31/12/2024, so the forecast period is 01/01/2025 -> 31/01/2025
forecast_window_start = pd.to_datetime("2025-01-01")
forecast_window_end   = pd.to_datetime("2025-01-31")

# Filter the daily data to only include the forecast period
mask = (df_daily['Date'] >= forecast_window_start) & (df_daily['Date'] <= forecast_window_end)
df_forecast_real = df_daily.loc[mask]

# Extract the 'Price' values (these are the real prices for the forecast period)
real_prices_next_30 = df_forecast_real['Price'].values

print("Real Price values for 01/01/2025 -> 31/01/2025:")
print(real_prices_next_30)